# Programming Examples

This repo contains examples of python and r code created by Nicholas Gleadall.
Below is a brief explanation of each tool.  

NGS_data_analysis
=================
<body>
<b>bam_ref_find.py</b><BR>

Usage = $ python bam_ref_find.py "bamfile"

Script which streams a .bamfile, 'soft clips' reads below the
desired quality and then writes the remaining reads to a new HQ bamfile.

It then streams the new HQ bamfile, building a dictionary of reference ID' and
counts how many reads are mapped to each one. It then returns the translated name
of most "mapped to" reference strain.

This tool can be easily adapted for many uses e.g. selection of most appropriate alternate sequences
in the latest human genome build HG38/GRCh38.

<b> reservoir_sample_fastq.py </b><BR>

Usage = $ python reservoir_sample_fastq.py "sample name" "number of reads desired"

Streams reads from a pair of fastq files (produced by Illumina sequencing), uses a reservoir sampling technique
to write a set number of random read pairs to two new fastq files.

This tool is for subsetting random reads from samples as required for bespoke analysis.

</body>

R_data_plotting
===============
<body>
<b>genome_coverage.R</b><BR>
Usage = $ Rscript genome_coverage.R "depth_file"

Simple plotting of data contained in .depth files generated by samtools.

"HIV_full_genome.png" is an example of the ouput, this shows number of reads (x) at a given genomic position (y).
y values can be edited as desired.

This was used as a quick and easy way of visualising pre sequencing, PCR amplicon coverage performance.

<b>Boxplot_freqs.R</b>

Example of an R analysis on more complex data. The aim was to determine the
reliability of mutation prevalences obtained from a HIV-1 sample by NGS.

The "Boxplot_example.png" is an example output of this analysis, it shows the mutations observed in 10 repeats
of a single sample and the prevelance at which each mutation was observed at.
</body>

analysis_pipelines
==================
